Add feature flags for Phase 2 (#13324)
WordPress/gutenberg: docs/designers-developers/developers/feature-flags.md

# Feature Flags

With phase 2 of the Gutenberg project there's a need for improved control over how code changes are released. Newer features developed for phase 2 and beyond should only be released to the Gutenberg plugin, while improvements and bug fixes should still continue to make their way into core releases.

The technique for handling this is known as a 'feature flag'. 

> ## Introducing `window.GUTENBERG_PHASE`

Reviewer: While it's quite thorough, I think most people coming to this document care most about the **Basic Use** section, so at the least I might suggest moving it to precede this section, if not simplify / cut parts of this section.

The `window.GUTENBERG_PHASE` constant is a global variable containing a number representing the phase. When the codebase is built for the plugin, this variable will be set to `2`. When building for core, it will be set to `1`.

### How it works

> More precisely, the `GUTENBERG_PHASE` variable isn't actually set to a value. Instead, any instances of `window.GUTENBERG_PHASE` in the codebase will be replaced during the webpack build by webpack's define plugin (https://webpack.js.org/plugins/define-plugin/).

Reviewer 2: Simplify explanation:

```suggestion
During the webpack build, any instances of `window.GUTENBERG_PHASE` will be replaced using webpack's define plugin (https://webpack.js.org/plugins/define-plugin/).
```

 ```js
 phaseTwoFeature();
 ```

Similarly when building for core, the condition in the following `if` statement always resolves to false:
```js
if ( 1 === 2 ) {
	phaseTwoFeature();
}
```

The minification process will remove the entire `if` statement including the body, ensuring code destined for phase 2 is not included in the built JavaScript intended for core.

## Basic Use

A phase 2 function or constant should be exported using the following ternary syntax:

```js
function myPhaseTwoFeature() {
	// implementation
}

export const phaseTwoFeature = window.GUTENBERG_PHASE === 2 ? myPhaseTwoFeature : undefined;
```

In phase 1 environments the `phaseTwoFeature` export will be `undefined`.

If you're attempting to import and call a phase 2 feature, be sure to wrap the call to the function in an if statement to avoid an error:
```js
import { phaseTwoFeature } from '@wordpress/foo';

if ( window.GUTENBERG_PHASE === 2) {
	phaseTwoFeature();
}
```

## FAQ

#### Since `GUTENBERG_PHASE` is a global object, why do I have to access it through the `window` object, e.g. `window.GUTENBERG_PHASE`?

> Third-party consumers of WordPress npm packages may have their own build process, resulting in `window.GUTENBERG_PHASE` not being replaced at build time with the appropriate number. In JavaScript, when attempting to access an undefined global variable an error is thrown, a situation that would cause lots of issues for third-parties (or additional build configuration). Accessing the `GUTENBERG_PHASE` variable through the `window` object circumvents this, the value of the variable evaluates to `undefined` resulting in WordPress packages continuing to work for third-parties.

Reviewer: This also hints at another problem though: They might have their own build process, so `window` can't be assumed to be defined either.

Author: Hmm, hadn't considered that. I suppose that would be usage outside the browser. 🤔 

It's a shame that the condition might end up becoming quite verbose. I'd wanted to keep it as sort and simple as possible.
